contracts/SignerBondsUniswapV2.sol

```diff
@@ -0,0 +1,254 @@
+        uint256 amountOutMin,
+        address[] calldata path,
+        address to,
+        uint256 deadline
+    ) external payable returns (uint256[] memory amounts);
+
+    function getAmountsOut(uint256 amountIn, address[] calldata path)
+        external
+        view
+        returns (uint256[] memory amounts);
+}
+
+/// @notice Interface for the Uniswap v2 pair.
+/// @dev This is an interface with just a few function signatures of the
+///      pair contract. For more info and function description please see:
+///      https://uniswap.org/docs/v2/smart-contracts/pair
+interface IUniswapV2Pair {
+    function getReserves()
+        external
+        view
+        returns (
+            uint112 reserve0,
+            uint112 reserve1,
+            uint32 blockTimestampLast
+        );
+}
+
+/// @title SignerBondsUniswapV2
+/// @notice ETH purchased by the risk manager from tBTC signer bonds needs to be
+///         swapped and deposited back to the coverage pool as collateral.
+///         SignerBondsUniswapV2 is a swap strategy implementation allowing the
+///         risk manager to store purchased ETH signer bonds so that any
+///         interested part can later swap them on Uniswap v2 exchange and
+///         deposit as coverage pool collateral. The governance can set crucial
+///         swap parameters: max allowed percentage impact, slippage tolerance
+///         and swap deadline, to force reasonable swap outcomes.
+contract SignerBondsUniswapV2 is ISignerBondsSwapStrategy, Ownable {
+    using SafeMath for uint256;
+
+    // One basis point is equivalent to 1/100th of a percent.
+    uint256 public constant BASIS_POINTS_DIVISOR = 10000;
+
+    IUniswapV2Router public uniswapRouter;
+    IUniswapV2Pair public uniswapPair;
+    address public targetTokenRecipient;
+    address public targetToken;
+
+    // Determines the maximum allowed price impact for the swap transaction.
+    // If transaction's price impact is higher, transaction will be reverted.
+    // Default value is 100 basis points (1%).
+    uint256 public maxAllowedPriceImpact = 100;
+    // Determines the slippage tolerance for the swap transaction.
+    // If transaction's slippage is higher, transaction will be reverted.
+    // Default value is 50 basis points (0.5%).
+    uint256 public slippageTolerance = 50;
+    // Determines the deadline in which the swap transaction has to be mined.
+    // If that deadline is exceeded, transaction will be reverted.
+    uint256 public swapDeadline = 20 minutes;
+
+    event UniswapV2SwapExecuted(uint256[] amounts);
+
+    constructor(IUniswapV2Router _uniswapRouter, CoveragePool _coveragePool) {
+        uniswapRouter = _uniswapRouter;
+        targetTokenRecipient = address(_coveragePool.assetPool());
+        targetToken = address(_coveragePool.collateralToken());
+        uniswapPair = IUniswapV2Pair(
+            computePairAddress(
+                _uniswapRouter.factory(),
+                _uniswapRouter.WETH(),
```

Comment:
We use `uniswapRouter.WETH()` in a couple of places in this contract and the address of `WETH` is constant. Do you think we can assign `address WETH = uniswapRouter.WETH()` to a contract level and then reuse it?

Reply:
Yep, I thought about this at the beginning. However, the gas reporter says this is a more expensive alternative. Contract construction will cost roughly `6500` more. Each `swapSignerBondsOnUniswapV2` will cost around `500-750` more gas, depending on the final implementation.

Reply:
Hm.. I see, yea, contract level vars are more expensive even if we just hardcode the WETH address. I'm okay what we have now.